### PR TITLE
Issue #6429: Add container page action icon in the browser toolbar

### LINKED ISSUE #6429: Add container icon in the browser toolbar

### PR DESCRIPTION
**Pull request by gabrielluong**
---

Fixes #6429

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/feature/toolbar/src/test/java/mozilla/components/feature/toolbar/ContainerToolbarFeatureTest.kt [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                    ),
+                    selectedTabId = "tab1",
+                    containers = mapOf(
+                        container.contextId to container
+                    )
+                )
+            )
+        )
+        val containerToolbarFeature = getContainerToolbarFeature(toolbar, store)
+        testDispatcher.advanceUntilIdle()
```
Comment (Amejia481) [inline components/feature/toolbar/src/test/java/mozilla/components/feature/toolbar/ContainerToolbarFeatureTest.kt:84]: Should this statement be after `store.dispatch(TabListAction.SelectTabAction("tab2")).joinBlocking()`?

File: components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ContainerToolbarAction.kt [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+import mozilla.components.support.utils.DrawableUtils.loadAndTintDrawable
+
+/**
+ * An action button that represents a container to be added to the toolbar.
+ *
+ * @param container Associated [ContainerState]'s icon and color to render in the toolbar.
+ * @param padding A optional custom padding.
+ * @param listener A optional callback that will be invoked whenever the button is pressed.
+ */
+open class ContainerToolbarAction(
```
Comment (Amejia481) [inline components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ContainerToolbarAction.kt:31]: Is there any specific reason why this class needs to be opened?

File: components/feature/toolbar/src/test/java/mozilla/components/feature/toolbar/ContainerToolbarTest.kt [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import mozilla.components.support.test.whenever
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertTrue
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.Mockito.spy
+import org.mockito.Mockito.verify
+
+@RunWith(AndroidJUnit4::class)
+class ContainerToolbarTest {
```
Comment (Amejia481) [inline components/feature/toolbar/src/test/java/mozilla/components/feature/toolbar/ContainerToolbarTest.kt:25]: nit: should this be renamed to `ContainerToolbarActionTest`?

File: components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ContainerToolbarFeature.kt [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            containerPageAction?.let {
+                toolbar.removePageAction(it)
+                toolbar.invalidateActions()
+                containerPageAction = null
+            }
+            return
+        } else if (containerState == containerPageAction?.container) {
+            // Do nothing since we're still in a tab with same container.
+            return
+        }
```
Comment (Amejia481) [inline components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ContainerToolbarFeature.kt:63]: nit: Maybe  the condition can be improve to a bit simpler.

From lines 52 to 57 we are validating that a normal tab doesn't have a page action container. But how that happens in the first place? Only tabs with containers must have page actions, right?

Comment (Amejia481) [reply components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ContainerToolbarFeature.kt]: 👍

File: components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ContainerToolbarAction.kt [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            .inflate(R.layout.mozac_feature_toolbar_container_action_layout, parent, false)
+
+        listener?.let { clickListener ->
+            rootView.setOnClickListener { clickListener.invoke() }
+        }
+
+        val backgroundResource =
+            parent.context.theme.resolveAttribute(android.R.attr.selectableItemBackgroundBorderless)
+
+        rootView.setBackgroundResource(backgroundResource)
```
[the remark this answers is not in the document]
Comment (Amejia481) [reply components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ContainerToolbarAction.kt:47]: If we are sure we are not doing overdraw, we can ignore the lint  `tools:ignore="Overdraw"`, as it just "Possible overdraw", we are doing this in other places, below an example :)

```xml
<TextView xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:tools="http://schemas.android.com/tools"
    android:id="@+id/mozac_browser_awesomebar_title"
    android:layout_width="wrap_content"
    android:layout_height="wrap_content"
    android:layout_margin="4dp"
    android:ellipsize="end"
    android:lines="1"
    android:padding="8dp"
    android:textSize="12sp"
    tools:ignore="Overdraw"
    tools:text="Fennec" />
```

**Comment by gabrielluong** [comment]
bors r=Amejia481

File: components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ContainerToolbarAction.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (gabrielluong) [reply components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ContainerToolbarAction.kt]: Fixed with tools:ignore="Overdraw"

**Pull request merged**